### PR TITLE
Add support for custom resolver with @inject

### PR DESCRIPTION
**Pull request by raymondfeng**
cc @bajtos @raymondfeng @ritch @superkhau

### CODE REVIEW

File: packages/context/src/inject.ts [hunk 1]
```diff
@@ -5,14 +5,26 @@
...
+
+/**
+ * Descriptor for an injection point
+ */
 export interface Injection {
-  bindingKey: string;
-  metadata?: { [attribute: string]: BoundValue; };
+  bindingKey: string; // Binding key
+  metadata?: { [attribute: string]: BoundValue; }; // Related metadata
+  resolve?: ResolverFunction; // A custom resolve function
```
Comment (superkhau) [inline packages/context/src/inject.ts]: nit: comments unnecessary (at least bindingKey // Binding Key), they will be described in the places where the actual injection is used

File: packages/context/src/inject.ts [hunk 2]
```diff
@@ -38,9 +50,10 @@ export interface Injection {
  * @param bindingKey What binding to use in order to resolve the value of the
  * decorated constructor parameter or property.
  * @param metadata Optional metadata to help the injection
+ * @param resolve Optional function to resolve the injection
  *
  */
-export function inject(bindingKey: string, metadata?: Object) {
+export function inject(bindingKey: string, metadata?: Object, resolve?: ResolverFunction) {
```
Comment (superkhau) [inline packages/context/src/inject.ts]: ie. here

File: packages/context/src/resolver.ts [hunk 3]
```diff
@@ -54,6 +54,21 @@ export function instantiateClass<T>(ctor: Constructor<T>, ctx: Context): T | Pro
...
+ * Resolve the value or promise for a given injection
+ * @param ctx Context
+ * @param injection Descriptor of the injection
+ */
+function resolve<T>(ctx: Context, injection: Injection): ValueOrPromise<T> {
+  if (injection.resolve) {
+    // A custom resolve function is provided
+    return injection.resolve(ctx, injection);
+  }
+  // Default to resolve the value from the context by binding key
```
Comment (superkhau) [inline packages/context/src/resolver.ts]: // Resolve the value from the context via the binding key by default
// Default to resolving the value from the context via the binding key

File: packages/context/src/resolver.ts [hunk 4]
```diff
@@ -54,6 +54,21 @@ export function instantiateClass<T>(ctor: Constructor<T>, ctx: Context): T | Pro
...
+  const binding = ctx.getBinding(injection.bindingKey);
+  return binding.getValue(ctx);
```
Comment (superkhau) [inline packages/context/src/resolver.ts]: return ctx.getBinding(injection.bindingKey).getValue(ctx);

Comment (superkhau) [inline packages/context/src/resolver.ts]:
```
// Default to resolving the value from the the context by the binding key if provided
return !injection.resolve ?
  ctx.getBinding(injection.bindingKey).getValue(ctx) :
  injection.resolve(ctx.injection)
```

File: packages/context/test/unit/resolver.test.ts [hunk 5]
```diff
@@ -23,14 +24,65 @@ describe('constructor injection', () => {
     const t = instantiateClass(TestClass, ctx) as TestClass;
     expect(t.foo).to.eql('FOO');
   });
+
+  it('can report error for missing binding key', () => {
```
Comment (superkhau) [inline packages/context/test/unit/resolver.test.ts]: nit: it('reports a resolution error when the binding key is missing

**Comment by raymondfeng** [comment]
For the record, another option is to use `metadata` to denote what resolver will be used and use `context` to find corresponding resolvers for a specific decorator type. For example,

1. Implement @repository as follows:
```
@repository(model, dataSource) ==> 
@inject('repository', {decoratorType: 'repository', model: model, dataSource: dataSource})
```

2. Register a custom resolver
```
ctx.bind('resolvers:repository').to(repositoryResolver);
```

3. Use the ctx to perform custom injection
```
const resolve = await ctx.get('resolvers:' + meta.decoratorType)
const repo = await resolve(ctx, injection);
```

**Pull request merged**